Make ColorMapper optional

TheE opened the pull request:
As explained in #2, `ColorMapper`'s dependency on janis is problematic as not every implementation of Bukkit bundles it. This pull request allows slf4bukkit to work on these platforms while maintaining existing functionality on all platforms that bundle janis:

- `ColorMapper`is changed from a static utility into a simple POJO that is initialized by using a factory method. This method can throw an Exception to indicate that `ColorMapper` cannot be initialized, e.g. because dependencies such as janis are missing.
- `BukkitLoggerAdapter` keeps a thread local reference to `ColorMapper` to allow possible multi-threading. If `ColorMapper`cannot be initialized, it is simply ignored and chat colors are not mapped to any equivalents (normally, if a platform does not bundle janis, it should be able to handle this mapping by itself).
- To make the build work, I also needed to register Spigot's repository in the pom.

TheE committed:
Use Spigot's Maven repository

TheE committed:
Only use ColorMapper if jansi is present, ignore it if not

This change allows slf4bukkit to run on Bukkit implementations without jansi by making ColorMapper optional: If it cannot be initialized, colors are not mapped. Instead this may (or may not) be done by the Bukkit implementation, that alos handles the logging system everything the output of slf4bukkit is to.

TheE committed:
Change group-id to 'io.github.mywarp' to prevent confusion

rjenkinsjr commented:
Thanks for the PR! I'm not familiar with other Bukkit implementations so I didn't realize that Jansi wasn't always available.

Rather than using an exception for flow control (which is costly and could be disorienting to others viewing the source code), I suggest that `ColorMapper` be made into an abstract class containing your factory `create()` method and making the `map()` method abstract. Two concrete subclasses would exist - `NotSupportedColorMapper` (which would follow the null object pattern and thus be no-op, but still coerce null to empty) and `DefaultColorMapper` (which would do what `ColorMapper` does now). The `create()` method would work as-is and return the correct object type.

That way, `BukkitLoggerAdapter` can call `mapper.map()` and be ignorant of the response, as it should, and we're also avoiding creating/throwing exceptions that don't serve any purpose outside of this library.

Two questions:
1. What was your rationale for moving the color mapping logic into the `mapColors` method?
1. Why are you using a `ThreadLocal` to create the `ColorMapper` instance? I'm not familiar with `ThreadLocal` so it's not obvious to me why one would want to do that.

One more thing:

> sexting functionality

I don't recall adding any of that to this library. Should I be expecting another PR? :laughing:

rjenkinsjr commented:
Also, regarding your `pom.xml` edit - I'm going to try to work next week on redesigning the build process so it doesn't use my private build server anymore, so I suspect I'll be pushing changes to `pom.xml` that you'll want to grab before I accept this PR.

TheE commented:
>Rather than using an exception for flow control (which is costly and could be disorienting to others viewing the source code), I suggest that `ColorMapper` be made into an abstract class containing your factory `create()` method and making the `map()` method abstract. Two concrete subclasses would exist - `NotSupportedColorMapper` (which would follow the null object pattern and thus be no-op, but still coerce null to empty) and `DefaultColorMapper` (which would do what `ColorMapper` does now). The `create()` method would work as-is and return the correct object type.
That way, `BukkitLoggerAdapter` can call `mapper.map()` and be ignorant of the response, as it should, and we're also avoiding creating/throwing exceptions that don't serve any purpose outside of this library.

That actually sounds better. Will do once you have worked on the `pom.xml`.

> What was your rationale for moving the color mapping logic into the mapColors method?

I wanted to encapsulate the null checking from the rest of the method. Can/will be removed when I implement the solution you outlined above.

> Why are you using a ThreadLocal to create the ColorMapper instance? I'm not familiar with ThreadLocal so it's not obvious to me why one would want to do that.

The way it is currently used, `ColorMapper.map(String)` is not thread safe. When I briefly skipped through `BukkitLoggerAdapter`, this class looked like it could (in principle) be used by multiple threads.

`ThreadLocale` keeps exactly one instance to an object for each thread, so that each thread uses its own instance only. If `BukkitLoggerAdapter` would be used by multiple threads simultaneously, each thread would be using its own instance of `ColorMapper`, thus working around any thread safety issues.

I could have also synchronized `ColorMapper.map()` but this could slow down the logging, as a thread would need to wait for an other thread's `map()`to finish.

> I don't recall adding any of that to this library. Should I be expecting another PR? 😆

We'll see about that. (I should not write PRs messages when I am tired...)

TheE commented:
Any news?

rjenkinsjr commented:
@TheE my apologies, had a lot of things come up over the past few weeks. I'll be taking a look at this over the weekend to see how much of the prep work we discussed previously I can get done.

rjenkinsjr commented:
@TheE I've finished moving the build onto CircleCI, so I can now reliably cut new releases of this library. You'll see I cut versions 0.1.5 and 0.1.6 today; they don't have any functional changes.

You can now pull my changes from `master`; pretty sure we overlap only on `pom.xml`.

TheE committed:
Revert "Change group-id to 'io.github.mywarp' to prevent confusion"

This reverts commit aa00ff8bd129981c7cbd741889f66d1dfd83c2e6.

TheE committed:
Revert "Only use ColorMapper if jansi is present, ignore it if not"

This reverts commit 50c130303453a393b1a312b8d0ebce67b9cbb658.

TheE committed:
Revert "Use Spigot's Maven repository"

This reverts commit e68d229674d4f6853a2255a5d1d5542765db3aa8.

TheE committed:
Merge branch 'master' of https://github.com/rjenkinsjr/slf4bukkit

TheE committed:
Only use ColorMapper if jAnsi is present, ignore it if not

This change allows slf4bukkit to run on Bukkit implementations without jansi by making ColorMapper an interface. Instances are created by using a ColorMappingFactory: If the Bukkit implementation bundles jAnsi, an AnsiColorMapper is returned that maps ChatColors to their ansi equivalents; if the Bukkit implementation does not bundle jAnsi, a NotSupportedColorMapper is returned that returns strings as is.

TheE commented:
@rjenkinsjr I decided to undo all my previous changes, before implementing the new logic as discussed. The last commit is the one you might want to actually look at.

(edit) Seems like a test is failing. No idea why though.

rjenkinsjr commented on src/main/java/info/ronjenkins/slf4bukkit/AnsiColorMapper.java:
```diff
@@ -0,0 +1,80 @@
+ */
+final class AnsiColorMapper implements ColorMapper {
+
+  AnsiColorMapper() throws Throwable {
```
I don't think `throws Throwable` is necessary. Because the constructor is no-op, I think only subclasses of `Error` could occur, and `Error` is unchecked so it doesn't have to be declared in the method signature.

TheE replied:
Right, `Error` (or, for that matter, `NoClassDefFoundError`) would be sufficient.

I chose `Throwable` because I think that having a checked exception here is better in terms of code maintenance. Everybody who ever works with this constructor gets a IDE warning if a catch clause is omitted, which should encourage him to look into the class in more detail and stumble upon my comment. If this where just an unchecked exception, somebody who does not dig into the source code would not get this kind of warning.

Basically the problem is, that in terms of Java in general, `NoClassDefFoundError` is an abdominal condition that should never occur or be catched. In this case however it is a 'normal' condition that should be expected - thus a checked exception fits better.

rjenkinsjr replied:
I see your point; it is useful to leverage the compiler to remind a developer that JAnsi could be missing at runtime. With that said, your existing comment needs to be converted into a Javadoc, for the same reason.

Given that the constructor is package-private (as it should be), I think saying `throws Throwable` is fine. Only people hacking directly on SLF4Bukkit would encounter this; someone merely using SLF4Bukkit as designed doesn't interact with the code at all, so this is merely an implementation detail.

TheE replied:
With the commit below, the comment is replaced by a detailed javadoc.

rjenkinsjr commented on src/main/java/info/ronjenkins/slf4bukkit/ColorMapperFactory.java:
```diff
@@ -0,0 +1,23 @@
+  public static ColorMapper create() {
+    try {
+      return new AnsiColorMapper();
+    } catch (Throwable throwable) {
```
If my comment on `AnsiColorMapper` is correct, then this should catch `Error`s and not `Throwable`s.

rjenkinsjr reviewed (commented):
@TheE pretty sure the CircleCI failure is due to how I'm handling env vars for releases, so that can be ignored. See my comments below.

Also, the JAnsi dependency declaration in `pom.xml` needs to be made `optional`.

TheE committed:
Replace comment with javadoc

rjenkinsjr commented:
@TheE looks good. Please update `pom.xml` per my previous comment (GitHub doesn't send notifications for edited comments, I guess). I'll work this weekend on getting this merged and making a new release.

TheE committed:
Make jansi an optional dependency in pom.xml

rjenkinsjr reviewed (approved):
I'm working this evening on merging these changes in and releasing version 1.0.0. May take a day or two. Thanks again!
